// impala/esm5/bmap.es5.js
/**
 * @license
 * Copyright Google LLC All Rights Reserved.
 *
 * Use of this source code is governed by an MIT-style license that can be
 * found in the LICENSE file at https://angular.io/license
 */
import { Directive, ElementRef, EventEmitter, Inject, InjectionToken, Input, NgModule, NgZone, Output } from '@angular/core';
import { CommonModule } from '@angular/common';
import { __extends } from 'tslib';
import * as tslib_1 from 'tslib';
import { DomPortalHost } from '@angular/cdk/portal';

var BMAP_CONFIG = new InjectionToken('bmap.config');

var BMapDirective = (function () {
    /**
     * @param {?} mapConfig
     * @param {?} elementRef
     * @param {?} zone
     */
    function BMapDirective(mapConfig, elementRef, zone) {
        this.mapConfig = mapConfig;
        this.elementRef = elementRef;
        this.zone = zone;
        this.mapLoaded = new EventEmitter();
    }
    /**
     * @return {?}
     */
    BMapDirective.prototype.ngAfterViewInit = function () {
        var _this = this;
        // this.loader.load().then(() => {
        var /** @type {?} */ map = this.map = this.zone.runOutsideAngular(function () {
            return new BMap.Map(_this.elementRef.nativeElement, {
                enableMapClick: false
            });
        });
        var /** @type {?} */ center = this.mapConfig.center || [120, 30], /** @type {?} */ zoom = this.mapConfig.zoom || 10;
        if (2 !== center.length) {
            throw new Error("invalid lon lat");
        }
        map.centerAndZoom(new BMap.Point(center[0], center[1]), zoom);
        this.mapLoaded.emit(map);
        // });
    };
    /**
     * @param {?} changes
     * @return {?}
     */
    BMapDirective.prototype.ngOnChanges = function (changes) {
        if (this.map) {
            if (changes['mapZoom']) {
                this.map.setZoom(this.mapZoom);
            }
            if (changes['mapCenter']) {
                this.map.setCenter(this.mapCenter);
            }
        }
    };
    BMapDirective.decorators = [
        { type: Directive, args: [{
                    selector: '[bmap]',
                },] },
    ];
    /**
     * @nocollapse
     */
    BMapDirective.ctorParameters = function () { return [
        { type: undefined, decorators: [{ type: Inject, args: [BMAP_CONFIG,] },] },
        { type: ElementRef, },
        { type: NgZone, },
    ]; };
    BMapDirective.propDecorators = {
        'mapLoaded': [{ type: Output },],
        'mapCenter': [{ type: Input },],
        'mapZoom': [{ type: Input },],
    };
    return BMapDirective;
}());

var BMapModule = (function () {
    function BMapModule() {
    }
    /**
     * @param {?} config
     * @return {?}
     */
    BMapModule.forRoot = function (config) {
        return {
            ngModule: BMapModule,
            providers: [
                { provide: BMAP_CONFIG, useValue: config }
            ]
        };
    };
    BMapModule.decorators = [
        { type: NgModule, args: [{
                    declarations: [BMapDirective],
                    imports: [CommonModule],
                    exports: [BMapDirective],
                    providers: [],
                },] },
    ];
    /**
     * @nocollapse
     */
    BMapModule.ctorParameters = function () { return []; };
    return BMapModule;
}());

var GenericMarker = (function (_super) {
    __extends(GenericMarker, _super);
    /**
     * @param {?} point
     * @param {?} opts
     * @param {?} data
     */
    function GenericMarker(point, opts, data) {
        var _this = _super.call(this, point, opts) || this;
        _this.data = data;
        return _this;
    }
    return GenericMarker;
}(BMap.Marker));

/**
 *
 */
var PortalHostOverlay = (function (_super) {
    __extends(PortalHostOverlay, _super);
    /**
     * @param {?} componentFactoryResolver
     * @param {?} appRef
     * @param {?} defaultInjector
     */
    function PortalHostOverlay(componentFactoryResolver, appRef, defaultInjector) {
        var _this = _super.call(this) || this;
        _this.hostDomElement = document.createElement("div");
        _this.hostDomElement.style.position = "absolute";
        _this.portalHost = new DomPortalHost(_this.hostDomElement, componentFactoryResolver, appRef, defaultInjector);
        return _this;
    }
    /**
     * @param {?} portal
     * @param {?} lon
     * @param {?} lat
     * @return {?}
     */
    PortalHostOverlay.prototype.attach = function (portal, lon, lat) {
        this.portalHost.attach(portal);
        // [lon, lat] = wgs84tobd09(lon, lat);
        this.point = new BMap.Point(lon, lat);
        this.draw();
    };
    /**
     * @return {?}
     */
    PortalHostOverlay.prototype.detach = function () {
        this.portalHost.detach();
    };
    /**
     * @param {?} map
     * @return {?}
     */
    PortalHostOverlay.prototype.initialize = function (map) {
        this.map = map;
        var /** @type {?} */ panes = map.getPanes();
        var /** @type {?} */ floatPane = panes.floatPane;
        if (floatPane) {
            floatPane.appendChild(this.hostDomElement);
        }
        return this.hostDomElement;
    };
    /**
     * @return {?}
     */
    PortalHostOverlay.prototype.isVisible = function () {
        return true;
    };
    /**
     * @return {?}
     */
    PortalHostOverlay.prototype.draw = function () {
        // console.log('PortalHostOverlay::draw()');
        if (!this.point) {
            return;
        }
        var /** @type {?} */ position = this.map.pointToOverlayPixel(this.point);
        this.hostDomElement.style.left = position.x /*- this._size / 2 */ + "px";
        this.hostDomElement.style.top = position.y /*- this._size / 2 */ + "px";
    };
    /**
     * @return {?}
     */
    PortalHostOverlay.prototype.show = function () {
        this.hostDomElement.style.display = "";
    };
    /**
     * @return {?}
     */
    PortalHostOverlay.prototype.hide = function () {
        this.hostDomElement.style.display = "none";
    };
    return PortalHostOverlay;
}(BMap.Overlay));

/**
 * 经纬度坐标转百度地图坐标
 * @param {?} lon
 * @param {?} lat
 * @return {?}
 */
function wgs84tobd09(lon, lat) {
    //     [lon, lat] = (<any>coordtransform).wgs84togcj02(lon, lat);
    //     [lon, lat] = (<any>coordtransform).gcj02tobd09(lon, lat);
    return [lon, lat];
}

/**
 * Generated bundle index. Do not edit.
 */

export { BMapDirective, BMapModule, BMAP_CONFIG, GenericMarker, PortalHostOverlay, wgs84tobd09 };
